packages/eslint-plugin/tests/rules/no-implied-eval.test.ts

```diff
@@ -0,0 +1,640 @@
+      `,
+    },
+    {
+      code: `
+const foo = {
+  fn() {},
+}
+
+setTimeout(foo.fn, 0);
+setInterval(foo.fn, 0);
+setImmediate(foo.fn);
+execScript(foo.fn);
+      `,
+    },
+    {
+      code: `
+const foo = {
+  fn: () => {},
+}
+const fn = 'fn';
+
+setTimeout(foo[fn], 0);
+setInterval(foo[fn], 0);
+setImmediate(foo[fn]);
+execScript(foo[fn]);
+      `,
+    },
+    {
+      code: `
+const foo = {
+  fn: () => {},
+}
+
+setTimeout(foo['fn'], 0);
+setInterval(foo['fn'], 0);
+setImmediate(foo['fn']);
+execScript(foo['fn']);
+      `,
+    },
+    {
+      code: `
+const foo: () => void = () => {
+};
+
+setTimeout(foo, 0);
+setInterval(foo, 0);
+setImmediate(foo);
+execScript(foo);
+      `,
+    },
+    {
+      code: `
+const foo: (() => () => void) = () => {
+  return () => {};
+}
+
+setTimeout(foo(), 0);
+setInterval(foo(), 0);
+setImmediate(foo());
+execScript(foo());
+      `,
+    },
+    {
+      code: `
+const foo: (() => () => void) = () => () => {};
+
+setTimeout(foo(), 0);
+setInterval(foo(), 0);
+setImmediate(foo());
+execScript(foo());
+      `,
+    },
+    {
+      code: `
+const foo = () => () => {};
+
+setTimeout(foo(), 0);
+setInterval(foo(), 0);
+setImmediate(foo());
+execScript(foo());
+      `,
+    },
+    {
+      code: `
+const foo = function foo () {
+  return function foo() {}
+}
+
+setTimeout(foo(), 0);
+setInterval(foo(), 0);
+setImmediate(foo());
+execScript(foo());
+      `,
+    },
+    {
+      code: `
+const foo = function () {
+  return function () {
+    return '';
+  }
+}
+
+setTimeout(foo(), 0);
+setInterval(foo(), 0);
+setImmediate(foo());
+execScript(foo());
+      `,
+    },
+    {
+      code: `
+const foo: (() => () => void) = function foo () {
+  return function foo() {}
+}
+
+setTimeout(foo(), 0);
+setInterval(foo(), 0);
+setImmediate(foo());
+execScript(foo());
+      `,
+    },
+    {
+      code: `
+function foo() {
+  return function foo() {
+    return () => {};
+  };
+}
+
+setTimeout(foo()(), 0);
+setInterval(foo()(), 0);
+setImmediate(foo()());
+execScript(foo()());
+      `,
+    },
+    {
+      code: `
+class Foo {
+  static fn = () => {};
+}
+
+setTimeout(Foo.fn, 0);
+setInterval(Foo.fn, 0);
+setImmediate(Foo.fn);
+execScript(Foo.fn);
+      `,
+    },
+    {
+      code: `
+class Foo {
+  fn() {}
+}
+
+const foo = new Foo();
+
+setTimeout(foo.fn, 0);
+setInterval(foo.fn, 0);
+setImmediate(foo.fn);
+execScript(foo.fn);
+      `,
+    },
+    {
+      code: `
+class Foo {
+  fn() {}
+}
+const foo = new Foo();
+const fn = foo.fn;
+
+setTimeout(fn.bind(null), 0);
+setInterval(fn.bind(null), 0);
+setImmediate(fn.bind(null));
+execScript(fn.bind(null));
+      `,
+    },
+    {
+      code: `
+const fn = (foo: () => void) => {
+  setTimeout(foo, 0);
+  setInterval(foo, 0);
+  setImmediate(foo);
+  execScript(foo);
+}
+      `,
+    },
+  ],
+
+  invalid: [
+    {
+      code: `
+setTimeout('x = 1', 0);
+setInterval('x = 1', 0);
+setImmediate('x = 1');
+execScript('x = 1');
```

Comment:
Question:
Do we want to have handling for these cases? Or are we okay with just handling the normal call?
These cases are getting into a bit of the world of "someone's trying really hard to do something weird", so it's probably okay to ignore them.

```ts
setTimeout.call(<something>, 'x = 1', 0);
setInterval.call(<something>, 'x = 1', 0);
setImmediate.call(<something>, 'x = 1');
execScript.call(<something>, 'x = 1');
```
```ts
setTimeout.apply(<something>, ['x = 1', 0]);
setInterval.apply(<something>, ['x = 1', 0]);
setImmediate.apply(<something>, ['x = 1']);
execScript.apply(<something>, ['x = 1']);
```